ml-agents/mlagents/trainers/environment_parameter_manager.py

```diff
@@ -0,0 +1,142 @@
+from typing import Dict, List, Tuple
+import numpy as np
+import math
+from mlagents.trainers.settings import (
+    EnvironmentParameterSettings,
+    CompletionCriteriaSettings,
+    ParameterRandomizationSettings,
+)
+from mlagents.trainers.training_status import GlobalTrainingStatus, StatusType
+
+from mlagents_envs.logging_util import get_logger
+
+logger = get_logger(__name__)
+
+
+class EnvironmentParameterManager:
+    def __init__(
+        self,
+        settings: Dict[str, EnvironmentParameterSettings],
+        run_seed: int,
+        restore: bool,
+    ):
+        self._dict_settings = settings
+        for parameter_name in self._dict_settings.keys():
+            initial_lesson = GlobalTrainingStatus.get_parameter_state(
+                parameter_name, StatusType.LESSON_NUM
+            )
+            if initial_lesson is None or not restore:
+                GlobalTrainingStatus.set_parameter_state(
+                    parameter_name, StatusType.LESSON_NUM, 0
+                )
+        self._smoothing_values: Dict[str, float] = {}
```

Comment:
`defaultdict(float)` ?

Reply:
Rename to `_smoothed_values`, or something to convey that it's an exponential moving average. "smoothing" makes it sound like it controls the interpolation amount (which is actually hardcoded to .25).

Reply:
Is there a reason for using defaultdict(float) over {} ?

Reply:
I don't feel that strongly about the defaultdict, in this case it only saves a few lines (since we know all the key ahead of time).